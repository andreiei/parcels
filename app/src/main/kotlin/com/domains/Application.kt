package com.domains

import com.domains.DomainsInventory.Companion.DomainsCount
import com.domains.models.Email

class Application(
    private val domainsInventory: DomainsInventory = DomainsInventory(),
) {
    fun start() {
        println("Enter one email address per line (type 'e' to finish)")

        while (true) {
            val line: String = readlnOrNull()?.trim() ?: break
            if (line.contains(' ')) {
                System.err.println("Input contains spaces, skipping: '$line'")
                continue
            }
            if (line.equals("e", ignoreCase = true)) break

            when (val email: Email? = Email.from(line)) {
                null -> System.err.println("Invalid email: '$line' â€” skipping.")
                else -> email.getDomain()?.let { domainsInventory.add(it) }
            }
        }

        val foundDomains: List<DomainsCount> = domainsInventory.get(10)
        printOutDomains(foundDomains)
    }

    private fun printOutDomains(foundDomains: List<DomainsCount>) {
        when (foundDomains.size) {
            0 -> ("No email domains found.").also { return }
            else -> {
                println("\nTop ${foundDomains.size} domains:")
                foundDomains.forEach { domain ->
                    println("${domain.domain.value} ${domain.count}")
                }
            }
        }
    }
}
