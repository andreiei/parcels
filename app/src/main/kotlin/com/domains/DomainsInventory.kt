package com.domains

import com.domains.models.Domain

/**
 * DomainsInventory is a collection that tracks the occurrence of domains.
 */
class DomainsInventory {
    private val domains: MutableMap<Domain, Int> = mutableMapOf()

    /**
     * Adds a domain to the collection, incrementing its count.
     */
    fun add(domain: Domain) {
        domains[domain] = domains.getOrDefault(domain, 0) + 1
    }

    /**
     * Returns a list of the top N domains sorted by their count in descending order.
     * @param n The number of top domains to return. Defaults to 10.
     */
    fun get(n: Int = 10): List<DomainsCount> {
        return domains.entries
            .sortedByDescending { it.value }
            .take(n)
            .map { DomainsCount(it.key, it.value) }
    }

    companion object {
        data class DomainsCount(
            val domain: Domain,
            val count: Int,
        )
    }
}
