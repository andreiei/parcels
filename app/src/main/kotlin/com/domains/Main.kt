package com.domains

fun main() {
    val application = Application()
    application.start()
}
