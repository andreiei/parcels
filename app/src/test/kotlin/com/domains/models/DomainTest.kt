package com.domains.models

import com.domains.shouldMatchSnapshot
import io.kotest.core.spec.style.FreeSpec

class DomainTest : FreeSpec({
    "should validate domain" {
        listOf(
            "domain.com",
            "subdomain.domain.com",
            "sub-domain.domain.com",
            "domain",
            "@domain",
            "domain..com",
            "peter@-domain.com",
        )
            .associateWith { domain ->
                when (Domain.from(domain)?.value) {
                    null -> "Invalid"
                    else -> "valid"
                }
            }
            .shouldMatchSnapshot(this)
    }
})
