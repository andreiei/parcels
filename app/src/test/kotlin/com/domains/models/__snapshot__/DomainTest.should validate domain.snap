{
    "domain.com": "valid",
    "subdomain.domain.com": "valid",
    "sub-domain.domain.com": "valid",
    "domain": "Invalid",
    "@domain": "Invalid",
    "domain..com": "Invalid",
    "peter@-domain.com": "Invalid"
}
